### PR TITLE
feat (worker, api): workflow run [wip]

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/workflow/execute_node_run.go [hunk 1]
```diff
@@ -63,33 +62,38 @@ func execute(db gorp.SqlExecutor, n *sdk.WorkflowNodeRun) error {
...
 				return errSync
 			}
 			if end {
-				log.Debug("workflow.execute> Begin [#%d.%d] runID=%d. Node is %s", n.Number, n.SubNumber, n.WorkflowRunID, newStatus)
-				//The job is over
-				//Delete the line in workflow_node_run_job
-				if err := DeleteNodeJobRuns(db, n.ID); err != nil {
-					return sdk.WrapError(err, "workflow.execute> Unable to delete node %d job runs ", n.ID)
-				}
+				//log.Debug("workflow.execute> Stage (%d-%s) [#%d.%d] ends runID=%d. Node is %s - stage is %s", stageIndex, stage.Name, n.Number, n.SubNumber, n.WorkflowRunID, newStatus, stage.Status.String())
```
Comment (yesnault) [inline engine/api/workflow/execute_node_run.go]: remove log or comment ?

File: engine/api/workflow/run_workflow_test.go [hunk 2]
```diff
@@ -89,23 +89,24 @@ func TestManualRun1(t *testing.T) {
...
 	test.NoError(t, err)
 
-	wr, err := ManualRun(db, w1, &sdk.WorkflowNodeRunManual{
+	_, err = ManualRun(db, w1, &sdk.WorkflowNodeRunManual{
 		User: *u,
 	})
 	test.NoError(t, err)
 
-	m, _ := dump.ToMap(wr)
+	/*
```
Comment (yesnault) [inline engine/api/workflow/run_workflow_test.go]: remove block?

File: engine/api/workflow/run_workflow_test.go [hunk 3]
```diff
@@ -133,16 +134,17 @@ func TestManualRun1(t *testing.T) {
...
-	keys = []string{}
-	for k := range m {
-		keys = append(keys, k)
-	}
-	sort.Strings(keys)
-	for _, k := range keys {
-		t.Logf("%s: \t%s", k, m[k])
-	}
-
+	/*
```
Comment (yesnault) [inline engine/api/workflow/run_workflow_test.go]: remove block ?

File: engine/worker/builtin_artifact.go [hunk 4]
```diff
@@ -104,24 +202,49 @@ func runArtifactDownload(*currentWorker) BuiltInAction {
...
 		tag = url.QueryEscape(tag)
 
-		if pipeline == "" {
+		sendLog(fmt.Sprintf("Downloading artifacts from into '%s'...", path))
+
+		n, err := strconv.ParseInt(number, 10, 64)
+		if err != nil {
 			res.Status = sdk.StatusFail.String()
-			res.Reason = fmt.Sprintf("pipeline variable is empty. aborting\n")
+			res.Reason = fmt.Sprintf("cds.run.numer variable is not valid. aborting\n")
```
Comment (yesnault) [inline engine/worker/builtin_artifact.go]: cds.run.number

File: sdk/cdsclient/client_queue.go [hunk 5]
```diff
@@ -119,3 +126,65 @@ func (c *client) QueueSendResult(id int64, res sdk.Result) error {
 	}
 	return nil
 }
+
+func (c *client) QueueArtifactUpload(id int64, tag, filePath string) error {
```
Comment (yesnault) [inline sdk/cdsclient/client_queue.go]: doc

File: sdk/cdsclient/client_queue.go [hunk 6]
```diff
@@ -119,3 +126,65 @@ func (c *client) QueueSendResult(id int64, res sdk.Result) error {
...
+	for i := 0; i <= c.config.Retry; i++ {
+		var code int
+		_, code, err = c.UploadMultiPart("POST", uri, body, SetHeader("Content-Disposition", "attachment; filename="+name), SetHeader("Content-Type", writer.FormDataContentType()))
+		if err == nil || code < 300 {
+			return nil
+		}
+		time.Sleep(1 * time.Second)
+	}
+
+	return fmt.Errorf("x5: %s", err)
```
Comment (yesnault) [inline sdk/cdsclient/client_queue.go]: x5 -> x%s, c.config.Retry

File: sdk/cdsclient/client_workflow.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"io"
+
+	"fmt"
+
+	"github.com/ovh/cds/sdk"
+)
+
+func (c *client) WorkflowRun(projectKey string, name string, number int64) (*sdk.WorkflowRun, error) {
```
Comment (yesnault) [inline sdk/cdsclient/client_workflow.go]: doc

File: sdk/cdsclient/client_workflow.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	url := fmt.Sprintf("/project/%s/workflows/%s/runs/%d", projectKey, name, number)
+	run := sdk.WorkflowRun{}
+	if _, err := c.GetJSON(url, &run); err != nil {
+		return nil, err
+	}
+	return &run, nil
+}
+
+func (c *client) WorkflowRunArtifacts(projectKey string, name string, number int64) ([]sdk.Artifact, error) {
```
Comment (yesnault) [inline sdk/cdsclient/client_workflow.go]: doc

File: sdk/cdsclient/client_workflow.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+	url := fmt.Sprintf("/project/%s/workflows/%s/runs/%d/artifacts", projectKey, name, number)
+	arts := []sdk.Artifact{}
+	if _, err := c.GetJSON(url, &arts); err != nil {
+		return nil, err
+	}
+	return arts, nil
+}
+
+func (c *client) WorkflowNodeRun(projectKey string, name string, number int64, nodeRunID int64) (*sdk.WorkflowNodeRun, error) {
```
Comment (yesnault) [inline sdk/cdsclient/client_workflow.go]: doc

File: sdk/cdsclient/client_workflow.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	url := fmt.Sprintf("/project/%s/workflows/%s/runs/%d/nodes/%d", projectKey, name, number, nodeRunID)
+	run := sdk.WorkflowNodeRun{}
+	if _, err := c.GetJSON(url, &run); err != nil {
+		return nil, err
+	}
+	return &run, nil
+}
+
+func (c *client) WorkflowNodeRunArtifacts(projectKey string, name string, number int64, nodeRunID int64) ([]sdk.Artifact, error) {
```
Comment (yesnault) [inline sdk/cdsclient/client_workflow.go]: doc

File: sdk/cdsclient/client_workflow.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	url := fmt.Sprintf("/project/%s/workflows/%s/runs/%d/nodes/%d/artifacts", projectKey, name, number, nodeRunID)
+	arts := []sdk.Artifact{}
+	if _, err := c.GetJSON(url, &arts); err != nil {
+		return nil, err
+	}
+	return arts, nil
+}
+
+func (c *client) WorkflowNodeRunArtifactDownload(projectKey string, name string, artifactID int64, w io.Writer) error {
```
Comment (yesnault) [inline sdk/cdsclient/client_workflow.go]: doc

File: sdk/workflow.go [hunk 12]
```diff
@@ -183,6 +183,26 @@ type WorkflowNode struct {
 	Triggers         []WorkflowNodeTrigger `json:"triggers,omitempty" db:"-"`
 }
 
+//EqualsTo returns true ia a node has the same pipeline, and context than another
```
Comment (yesnault) [inline sdk/workflow.go]: // EqualsTo returns true if a node has the same pipeline and context than another

File: sdk/cdsclient/client_queue.go (hunk 5 above)
Comment (fsamin) [inline sdk/cdsclient/client_queue.go]: nop

File: sdk/cdsclient/client_workflow.go (hunk 7 above)
Comment (fsamin) [inline sdk/cdsclient/client_workflow.go]: nope

File: sdk/cdsclient/client_workflow.go (hunk 8 above)
Comment (fsamin) [inline sdk/cdsclient/client_workflow.go]: nope

File: sdk/cdsclient/client_workflow.go (hunk 10 above)
Comment (fsamin) [inline sdk/cdsclient/client_workflow.go]: nope

File: sdk/cdsclient/client_workflow.go (hunk 11 above)
Comment (fsamin) [inline sdk/cdsclient/client_workflow.go]: nope

File: sdk/cdsclient/client_workflow.go (hunk 9 above)
Comment (fsamin) [inline sdk/cdsclient/client_workflow.go]: nope

**Pull request merged**